### PR TITLE
The ExecutionResult.Data property is serialized only if the request was executed.

### PR DESCRIPTION
**Pull request by sungam3r**
See https://spec.graphql.org/June2018/#sec-Data
> If an error was encountered before execution begins, the data entry should not be present in the result.
> If an error was encountered during the execution that prevented a valid response, the data entry in the response should be null.

### CODE REVIEW

File: src/GraphQL/Execution/ExecutionResult.cs [hunk 1]
```diff
@@ -11,6 +11,13 @@ namespace GraphQL
     /// </summary>
     public class ExecutionResult
     {
+        /// <summary>
+        /// Indicates if the operation included execution. If an error was encountered BEFORE execution begins,
+        /// the data entry SHOULD NOT be present in the result. If an error was encountered DURING the execution
+        /// that prevented a valid response, the data entry in the response SHOULD BE null.
+        /// </summary>
+        public bool Executed { get; set; }
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionResult.cs:19]: @Shane32 If there is a better way to distinguish between the errors before and after execution stage, then suggest.

File: src/GraphQL/Execution/DocumentExecuter.cs [hunk 2]
```diff
@@ -266,24 +269,15 @@ await listener.AfterExecutionAsync(context)
                         }
                 }
 
-                if (context.Errors.Count > 0)
-                {
-                    result.Errors = context.Errors;
-                }
+                result.AddErrors(context.Errors);
```
Comment (Shane32) [inline src/GraphQL/Execution/DocumentExecuter.cs:272]: Why are we making these changes?  It just adds more allocations during execution.  Plus the code is harder to read.

File: src/GraphQL/Execution/DocumentExecuter.cs [hunk 3]
```diff
@@ -298,18 +292,14 @@ await listener.AfterExecutionAsync(context)
...
                 }
 
-                result = new ExecutionResult
-                {
-                    Errors = new ExecutionErrors
-                    {
-                        ex is ExecutionError executionError ? executionError : new UnhandledError(exceptionContext?.ErrorMessage ?? "Error executing document.", ex)
-                    }
-                };
+                (result ??= new ExecutionResult()).AddError(ex is ExecutionError executionError ? executionError : new UnhandledError(exceptionContext?.ErrorMessage ?? "Error executing document.", ex));
```
Comment (Shane32) [inline src/GraphQL/Execution/DocumentExecuter.cs:295]: And, this is now harder to read than it was before.

File: src/GraphQL/Execution/ExecutionResult.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionResult.cs:19]: This seems fine to me.

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 2 above)
Comment (Shane32) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: It's fine though.  Optimizing runtime when errors are present isn't a priority.

Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: Why more ?

Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: My thought was that errors from the context should complement the errors already recorded in the result, and not replace them. This is not true?

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 3 above)
Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:295]: My thought was that errors from the context should complement the errors already recorded in the result, and not replace them. This is not true? What if `result` is not null?

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 2 above)
Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: What if `result.Errors` is not null/empty?

Comment (Shane32) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: With `AddErrors`, it has to iterate over the errors in `context.Errors` and add them to a new instance within `result`.  So it is more allocations.

Comment (Shane32) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: > My thought was that errors from the context should complement the errors already recorded in the result, and not replace them. This is not true?

Makes sense, although there is no existing possibility of this occurring, unless perhaps a completely custom executer was used.

File: src/GraphQL/Execution/ExecutionResult.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL/Execution/ExecutionResult.cs:19]: What about default value? Both `true`/`false` options have disadvantages.

**Review by Shane32 — APPROVED** [review]

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 3 above)
Comment (Shane32) [reply src/GraphQL/Execution/DocumentExecuter.cs:295]: Ok.  Currently that is only possible if a document listener throws an error after an execution.

File: src/GraphQL/Execution/DocumentExecuter.cs [hunk 4]
```diff
@@ -230,6 +231,8 @@ await listener.AfterValidationAsync(context, validationResult)
                     };
                 }
 
+                executionOccurred = true;
```
Comment (Shane32) [inline src/GraphQL/Execution/DocumentExecuter.cs:234]: The positioning here seems to make sense.  The document listeners could throw an exception, but they could be considered part of the execution stage....

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 2 above)
Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: > It's fine though. Optimizing runtime when errors are present isn't a priority.

I have not optimized.

> With AddErrors, it has to iterate over the errors in context.Errors and add them to a new instance within result. So it is more allocations.

OK. Possible one `List` instance.

> Makes sense, although there is no existing possibility of this occurring, unless perhaps a completely custom executer was used.

I just didn't like that I saw the potential for losing errors. I made sure that they are not lost in any situation, without understanding whether it is possible now or in the future. Is it good?

Comment (Shane32) [reply src/GraphQL/Execution/DocumentExecuter.cs:272]: 👍

File: src/GraphQL/Execution/ExecutionResult.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionResult.cs:19]: `false`

**Comment by Shane32** [comment]
Good to merge.

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 4 above)
Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:234]: Yes. Listeners "belong" to execution.

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 3 above)
Comment (sungam3r) [reply src/GraphQL/Execution/DocumentExecuter.cs:295]: Indeed.

**Pull request merged**